File: project_euler/cpp/92.cpp
#include <iostream>

using namespace::std;

int next_chain(int num){
	int sum = 0;
	int digit;
	while(num > 0){
		digit = num % 10;
		sum += pow(digit, 2);
		num = floor(num/10);
	}

	return sum;
}


int find_termination(int start_num){
	int num = start_num;
	while((num != 1) && (num != 89)){ // using the correct "&&" operator made the program very fast
		num = next_chain(num);
	}

	return num;
}


int main(){
	int count_89 = 0;
	for(int num=1; num<10000000; num++){
		if(find_termination(num) == 89){
			count_89 += 1;
		}
	}

	cout << count_89;
}

File: project_euler/python/92.py
# chains ALWAYS end at 1 or 89 – use that

def next_chain(num):
    return sum(int(digit)**2 for digit in str(num))

def find_termination(start_num):
    num = start_num
    while (num != 1) and (num != 89): # this was the problem
        num = next_chain(num)
    return num


def find_arriving_nums():
    count_89 = 0
    for num in range(1,10000000):
        if find_termination(num) == 89:
            count_89 += 1
    return count_89



print(find_arriving_nums())
